common/security-features/subresource/subresource.py

```diff
@@ -58,107 +61,106 @@ def create_url(request,
                              parsed.scheme)
-        hostname = parsed.netloc.split(':')[0]
-        port = request.server.config["ports"][parsed.scheme][0]
-        destination_netloc = ":".join([hostname, str(port)])
+        hostname = parsed.netloc.split(u':')[0]
+        port = request.server.config[u"ports"][parsed.scheme][0]
+        destination_netloc = u":".join([hostname, str(port)])
 
     if swap_origin:
         destination_netloc = __get_swapped_origin_netloc(destination_netloc)
 
-    parsed_query = urlparse.parse_qsl(parsed.query, keep_blank_values=True)
-    parsed_query = filter(lambda x: x[0] != query_parameter_to_remove,
-                          parsed_query)
+    parsed_query = parse_qsl(parsed.query, keep_blank_values=True)
+    parsed_query = [x for x in parsed_query if x[0] != query_parameter_to_remove]
 
-    destination_url = urlparse.urlunsplit(urlparse.SplitResult(
+    destination_url = urlunsplit(SplitResult(
         scheme = scheme,
         netloc = destination_netloc,
         path = parsed.path,
-        query = urllib.urlencode(parsed_query),
+        query = urlencode(parsed_query),
         fragment = None))
 
     return destination_url
 
 
 def preprocess_redirection(request, response):
-    if "redirection" not in request.GET:
+    if b"redirection" not in request.GET:
         return False
 
-    redirection = request.GET["redirection"]
+    redirection = request.GET[b"redirection"]
 
-    if redirection == "no-redirect":
+    if redirection == b"no-redirect":
         return False
-    elif redirection == "keep-scheme":
+    elif redirection == b"keep-scheme":
         redirect_url = create_url(request, swap_scheme=False)
-    elif redirection == "swap-scheme":
+    elif redirection == b"swap-scheme":
         redirect_url = create_url(request, swap_scheme=True)
-    elif redirection == "downgrade":
+    elif redirection == b"downgrade":
         redirect_url = create_url(request, downgrade=True)
-    elif redirection == "keep-origin":
+    elif redirection == b"keep-origin":
         redirect_url = create_url(request, swap_origin=False)
-    elif redirection == "swap-origin":
+    elif redirection == b"swap-origin":
         redirect_url = create_url(request, swap_origin=True)
     else:
-        raise ValueError("Invalid redirection type '%s'" % redirection)
+        raise ValueError(u"Invalid redirection type '%s'" % isomorphic_decode(redirection))
 
     redirect(redirect_url, response)
     return True
 
 
 def preprocess_stash_action(request, response):
-    if "action" not in request.GET:
+    if b"action" not in request.GET:
         return False
 
-    action = request.GET["action"]
+    action = request.GET[b"action"]
 
-    key = request.GET["key"]
+    key = request.GET[b"key"]
     stash = request.server.stash
-    path = request.GET.get("path", request.url.split('?'))[0]
+    path = request.GET.get(b"path", request.url.split(u'?'))[0]
```

Comment:
It looks like we should call isomorphic_encode on the second argument here.

Also, it looks like a pre-existing bug but I'm pretty sure the [0] is meant to be applied to the result of split(), rather than the result of get().

Reply:
Yes, I missed this one :). Code updated.